cdc/sorter/leveldb/leveldb.go

```diff
@@ -0,0 +1,214 @@
+	option.WriteL0SlowdownTrigger = lcfg.WriteL0SlowdownTrigger
+	option.WriteL0PauseTrigger = lcfg.WriteL0PauseTrigger
+	option.ErrorIfExist = true
+	option.NoSync = true
+
+	dbDir := filepath.Join(cfg.SortDir, fmt.Sprintf("%04d", id))
+	err := retry.Do(ctx, func() error {
+		err1 := os.RemoveAll(dbDir)
+		if err1 != nil {
+			log.Warn("clean data dir fails",
+				zap.String("dir", dbDir), zap.Error(err1))
+		}
+		return err1
+	},
+		retry.WithBackoffBaseDelay(500), // 0.5s
+		retry.WithMaxTries(4))           // 2s in total.
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	db, err := leveldb.OpenFile(dbDir, &option)
+	return db, errors.Trace(err)
+}
+
+// LevelActor is a leveldb actor, it reads, writes and deletes key value pair
+// in its leveldb.
+type LevelActor struct {
+	id       actor.ID
+	db       *leveldb.DB
+	wb       *leveldb.Batch
+	wbSize   int
+	iterSema *semaphore.Weighted
+	closedWg *sync.WaitGroup
+
+	metricWriteDuration prometheus.Observer
+	metricWriteBytes    prometheus.Observer
+}
+
+var _ actor.Actor = (*LevelActor)(nil)
+
+// NewLevelDBActor returns a leveldb actor.
+func NewLevelDBActor(
+	ctx context.Context, id int, db *leveldb.DB, cfg *config.SorterConfig,
+	wg *sync.WaitGroup, captureAddr string,
+) (*LevelActor, actor.Mailbox, error) {
+	idTag := fmt.Sprint(id)
+	// Write batch size should be larger than block size to save CPU.
+	wbSize := cfg.LevelDB.BlockSize * 16
+	wb := leveldb.MakeBatch(wbSize)
+	// IterCount limits the total number of opened iterators to release leveldb
+	// resources (memtables and SST files) in time.
+	iterSema := semaphore.NewWeighted(int64(cfg.LevelDB.Concurrency))
+	mb := actor.NewMailbox(actor.ID(id), cfg.LevelDB.Concurrency)
+	wg.Add(1)
+	return &LevelActor{
+		id:       actor.ID(id),
+		db:       db,
+		wb:       wb,
+		iterSema: iterSema,
+		wbSize:   wbSize,
+		closedWg: wg,
+
+		metricWriteDuration: sorterWriteDurationHistogram.WithLabelValues(captureAddr, idTag),
+		metricWriteBytes:    sorterWriteBytesHistogram.WithLabelValues(captureAddr, idTag),
+	}, mb, nil
+}
+
+func (ldb *LevelActor) close(err error) {
+	log.Info("leveldb actor quit", zap.Uint64("ID", uint64(ldb.id)), zap.Error(err))
+	ldb.closedWg.Done()
+}
+
+func (ldb *LevelActor) maybeWrite(force bool) error {
+	bytes := len(ldb.wb.Dump())
+	if bytes >= ldb.wbSize || (force && bytes != 0) {
+		startTime := time.Now()
+		err := ldb.db.Write(ldb.wb, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		ldb.metricWriteDuration.Observe(time.Since(startTime).Seconds())
+		ldb.metricWriteBytes.Observe(float64(bytes))
+
+		// Reset write batch or reclaim memory if it grows too large.
+		if cap(ldb.wb.Dump()) <= 2*ldb.wbSize {
+			ldb.wb.Reset()
+		} else {
+			ldb.wb = leveldb.MakeBatch(ldb.wbSize)
+		}
+	}
+	return nil
+}
+
+// Poll implements actor.Actor.
+// It handles tasks by writing kv, deleting kv and taking iterators.
+func (ldb *LevelActor) Poll(ctx context.Context, tasks []actormsg.Message) bool {
```

Comment:
long func, could this separate into small ones or more comments

Reply:
Added some comments.